apollo-router/src/plugins/telemetry/config_new/conditional.rs

```diff
@@ -0,0 +1,501 @@
+use crate::plugins::telemetry::otlp::TelemetryDataKind;
+
+/// The state of the conditional.
+#[derive(Debug, Default)]
+pub(crate) enum State<T> {
+    /// The conditional has not been evaluated yet or no value has been set via selector.
+    #[default]
+    Pending,
+    /// The conditional has been evaluated and the value has been obtained.
+    Value(T),
+    /// The conditional has been evaluated and the value has been returned, no further processing should take place.
+    Returned,
+}
+
+impl<T> From<T> for State<T> {
+    fn from(value: T) -> Self {
+        State::Value(value)
+    }
+}
+
+/// Conditional is a stateful structure that may be called multiple times during the course of a request/response cycle.
+/// As each callback is called the underlying condition is updated. If the condition can eventually be evaluated then it returns
+/// Some(true|false) otherwise it returns None.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct Conditional<Att> {
+    pub(crate) selector: Att,
+    pub(crate) condition: Option<Arc<Mutex<Condition<Att>>>>,
+    pub(crate) value: Arc<Mutex<State<opentelemetry::Value>>>,
+}
+
+impl<T> JsonSchema for Conditional<T>
+where
+    T: JsonSchema,
+{
+    fn schema_name() -> String {
+        format!("conditional_attribute_{}", type_name::<T>())
+    }
+
+    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
+        // Add the condition and also allow string as a fallback.
+        let mut selector = gen.subschema_for::<HashMap<String, T>>();
+        if let Schema::Object(schema) = &mut selector {
+            if let Some(object) = &mut schema.object {
+                object
+                    .properties
+                    .insert("condition".to_string(), gen.subschema_for::<Condition<T>>());
+            }
+        }
+        Schema::Object(SchemaObject {
+            subschemas: Some(Box::new(SubschemaValidation {
+                one_of: Some(vec![selector, gen.subschema_for::<String>()]),
+                ..Default::default()
+            })),
+            ..Default::default()
+        })
+    }
+}
+
+impl<Att> DefaultForLevel for Conditional<Att>
+where
+    Att: DefaultForLevel,
+{
+    fn defaults_for_level(
+        &mut self,
+        requirement_level: DefaultAttributeRequirementLevel,
+        kind: TelemetryDataKind,
+    ) {
+        self.selector.defaults_for_level(requirement_level, kind);
+    }
+}
+
+impl<Att, Request, Response> Selector for Conditional<Att>
+where
+    Att: Selector<Request = Request, Response = Response>,
+{
+    type Request = Request;
+    type Response = Response;
+
+    fn on_request(&self, request: &Self::Request) -> Option<opentelemetry::Value> {
+        match &self.condition {
+            Some(condition) => match condition.lock().evaluate_request(request) {
+                None => {
+                    if let Some(value) = self.selector.on_request(request) {
+                        *self.value.lock() = value.into();
```

Comment:
we should be careful here, because matching on something that was locked means the lock will be held for the entire length of the match